kv/memberlist/memberlist_client.go

```diff
@@ -1184,7 +1184,10 @@ func (m *KV) mergeValueForKey(key string, incomingValue Mergeable, casVersion ui
 	m.storeMu.Lock()
 	defer m.storeMu.Unlock()
 
-	curr := m.store[key].Clone()
+	// Note that we do not take a copy of the store value here, it is modified in-place.
+	// This is safe because the entire function runs under the store lock; we do not return
+	// the full state anywhere as is done elsewhere (i.e. Get/WatchKey/CAS).
+	curr := m.store[key]
```

Comment:
I do see two side effects of this change, but I don't believe either to be a problem. 
1. When no change is detected, the state is not written back to the map. Without the Clone, the state in the map effectively **is** updated, however, the state by definition, should not have changed, so I do not see that this in reality is a concern.
2. When we do RemoveTombstones (further down), this now updates the state in-place. When the state changes, this happens anyway. However there is a case where we do not update the state if there are no changes, following removal of tombstones. This confused me a little as I think the state _should_ be updated, albeit not propagated (which will be the case, with this change).

Reply:
I agree with your explanation. Both cases should be fine.

Reply:
1. Suggest documenting a requirement on the `Mergeable` interface that `Merge()` not modify the LHS without returning a change.
2. If you do the `Clone()` above `RemoveTombstones()`, does that change things ?  I expect tombstone removal is extremely rare, so you get much the same benefit without worrying about this change.

Reply:
> When no change is detected, the state is not written back to the map. Without the Clone, the state in the map effectively is updated, however, the state by definition, should not have changed, so I do not see that this in reality is a concern.

I also don't see a concern here.

Reply:
>When we do RemoveTombstones (further down), this now updates the state in-place.

There's a comment saying:
```
// Remove tombstones from change too. If change turns out to be empty after this,
// we don't need to change local value either!
```

To my understanding, after this PR the local value is changed. If so, we should update the comment.

Reply:
> Suggest documenting a requirement on the Mergeable interface that Merge() not modify the LHS without returning a change.

That's fine but we need to allow in-place changes for the sake of token normalization (#77). However, logically the contents will not change, but the physical layout may change (tokens normalized). I'll add a comment to that effect.

> If you do the Clone() above RemoveTombstones(), does that change things ? I expect tombstone removal is extremely rare, so you get much the same benefit without worrying about this change.

Unfortunately it would then Clone() every time. We don't know if we will actually remove tombstones before calling RemoveTombstones().

Reply:
Comments updated.